src/Set.sol

```diff
@@ -122,6 +127,7 @@ library Set {
         if (numElements == 1) {
             setStorage.numElements = 0;
             setStorage.firstElement = address(0);
+            setStorage.metadata = metadata;
```

Comment:
When `numElements == 1` it means that there's only one element in the set, why is `metadata` re-written as it is? Shouldn't it be resetted to `0`? Otherwise, you have a dirty `metadata` state (non-empty) that is inconsistent with the set state (empty)

Reply:
I'd argue the set metadata is Independent of the set content. What one does with it depends on the business logic of the contract using the lib. If one wanted to reset the metadata when when set becomes empty, they could explicitly use a function dedicated for setting the metadata.

In our case, we use the metadata in the controllers set to store the last successful account status check timestamp. The fact that we no longer have a controller enabled, doesn't mean the timestamp of the last successful check is suddenly invalid.

Reply:
But the information is bound to a `controller` because each controller could have a different set of rules.
A user "valid" for `ControllerA` does not mean that is valid for `ControllerB`.

Let's say that a controller validation is called, metadata is updated, and then the controller is disabled. At this point, we have a non-empty `metadata`.

A new controller is added and metadata is still non-empty. The current value of `metadata` is not related to the controller that has being added, but anyone that use it could be fooled by this fact.

I could agree that `metadata` in the `Set` context is different, and probably it's fine like it is right now, but I'm not sure that how it's handled in the `EthereumVaultConnector` is correct.

Reply:
To be clear: I've not looked up yet at `EVK`, I'm just looking at the `metadata` field with only the EVC context and what it represents with that narrow view.

Reply:
Yes, the controllers can have different set of rules but it does not matter in this context. The check (the actual controller `checkAccountStatus` function call) is only executed if there's exactly one controller enabled. It means that at the timestamp recorded, the account was considered healthy.

You're right though that if check is deferred for the account, the metadata might be stale if someone reads it, because the new check hasn't been executed yet. I added a little bit more natspec to explain that and, depending on the situation, the consumer of this data might need to check if the check is deferred for this account. This is exactly what happens in the EVK.

https://github.com/euler-xyz/ethereum-vault-connector/pull/158/files